README.md
[![!PyPi version](https://img.shields.io/pypi/v/wagtail-redisearch.svg)](https://pypi.org/project/wagtail-redisearch/)
[![!Python versions](https://img.shields.io/pypi/pyversions/wagtail-redisearch.svg)](https://pypi.org/project/wagtail-redisearch/)
[![!CI/CD status](https://github.com/TommasoAmici/wagtail-redisearch/actions/workflows/build.yaml/badge.svg)](https://github.com/TommasoAmici/wagtail-redisearch/actions/workflows/build.yaml)
[![!Code coverage status](https://codecov.io/gh/TommasoAmici/wagtail-redisearch/branch/main/graph/badge.svg)](https://codecov.io/gh/TommasoAmici/wagtail-redisearch)
[![!Formatted with Black](https://img.shields.io/badge/code%20style-black-000000.svg)](https://github.com/psf/black)

# wagtail-redisearch

A Django app to use [RediSearch](https://redis.com/modules/redis-search/) as a search
backend in Wagtail.

## Requirements

- Django >= 3.2
- Wagtail >= 2.15
- RediSearch v2

## Usage

Install [`wagtail-redisearch`](https://pypi.org/project/wagtail-redisearch/) from PyPi.

In your `settings.py` add `"wagtail_redisearch"` to `INSTALLED_APPS` and configure your search backend as follows:

```py
INSTALLED_APPS = [
  "wagtail_redisearch",
  ...
]

WAGTAILSEARCH_BACKENDS = {
    "default": {
        "BACKEND": "wagtail_redisearch",
        # optional parameters
        "INDEX": "custom-index-name", # default: "wagtail"
        "HOST": "127.0.0.1",
        "PORT": 6379,
        # you can add any option here to be used when initializing
        # a Redis client with redis-py.
        # e.g.
        "retry_on_error": True
    }
}
```

For more information about what options you can pass to the Redis client, look at the [official documentation](https://redis.readthedocs.io/en/stable/connections.html#generic-client).

`wagtail-redisearch` implements the interfaces described in [Backends Rolling your own](https://docs.wagtail.org/en/stable/topics/search/backends.html#rolling-your-own), thus usage in Wagtail requires no special adjustment.

To configure search on your models, follow the [official Wagtail documentation](https://docs.wagtail.org/en/stable/topics/search/index.html).

## Contributing

To contribute to this project you'll need RediSearch installed and `poetry`.

```sh
# install dependencies
poetry install

# run tests
make test

# run tests with tox
make test-all

# to run a minimal Wagtail application to test things out
make run
```


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

This project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [0.4.0] - 2022-04-21

[4c4baba](4c4babaac36ca4246b3c3cec2114bcbcddaa0c6d)...[4125466](4125466ce53625d04b63061d40bbdea7cdd98255)

### Features

- Added support for RelatedFields ([40bc184](40bc184080e6fd331ea339325f2f31359167505d))

## [0.3.3] - 2022-04-20

[ae09850](ae09850d64e7560888251d434547aad1e661b041)...[4c4baba](4c4babaac36ca4246b3c3cec2114bcbcddaa0c6d)

### Bug Fixes

- Applying any ordering would result in empty search results ([e92e091](e92e091162dd9b9d41ee880e1697ca2e8dcab90c))

### Documentation

- Added instructions to install app in Django settings ([af040f8](af040f8f17d891e0251026fd0a1baf9a79c06bcb))

## [0.3.2] - 2022-04-12

[48f93de](48f93de2e765771100a2c340f23bb77afc5d369a)...[ae09850](ae09850d64e7560888251d434547aad1e661b041)

### Bug Fixes

- Specify prefix when creating index ([1328454](132845427b2935cb32b53f56eba47a9137bde9c6))
- Specify index type HASH when creating index ([5807d0a](5807d0a14b65467ae73bdbea0ff5f673fbe1d854))
- To store time as a numeric field it needs to be formatted as a number ([7f71b57](7f71b579c7df93109062a621d04a8ea2d39274f9))

### Styling

- Added isort configuration ([d01440e](d01440e142de381b2612f5daa3d3d7cafd6dadc5))

### Testing

- Added tests for get_redis_field ([09fb82e](09fb82ec648782282c72d405ac2d2fc761585bdc))
- Added fixtures and search tests ([856cec9](856cec96e7bdd37b09ff88bbc5293d2a606a2f86))
- Added autocomplete test ([ab46c81](ab46c81c1b5475b9f00134809f59fba4c718e463))

### Ci

- Always run new tests ([341bd93](341bd93f26a31af9a784ad41980e8fafb94f4165))
- Added redisearch service to test workflows ([bd3e2c7](bd3e2c7e7d81c6aaa87f359f5c01bec2a329738d))

## [0.3.1] - 2022-04-12

[cf919a6](cf919a65ddf992afd843b131172a99dd5dc814b1)...[48f93de](48f93de2e765771100a2c340f23bb77afc5d369a)

### Bug Fixes

- Add_items doesn't expect a fields argument ([09f5c4b](09f5c4b52aa67290671ca8e87bdce4d3f0953d2e))
- Error when converting date objects to timestamps ([0fa7b29](0fa7b29e9d00a102de427ec2102ccf2022386356))
- Error when converting float to redis type ([5ad1bf7](5ad1bf7544f3639a8212c5b82d1c590538305ab3))
- Error when converting time objects to redis type ([2957229](295722935e045efc3d7bab396dad31d7565ebc10))

### Ci

- Use provided cliff config when generating changelog ([caad0ad](caad0ad7032932a1e656c5d1e91ad38e4d003a55))

## [0.3.0] - 2022-04-11

[29d0bf2](29d0bf24638df1ba53bea75b121f390190e6d431)...[cf919a6](cf919a65ddf992afd843b131172a99dd5dc814b1)

### Features

- Added support for more Django field types ([9b3d2f0](9b3d2f0b3d19062f3c28dfc8e1f5320e4a17a898))

### Miscellaneous Tasks

- Added badges to README ([8babce8](8babce8ccf1439abf2d9cddc92981d3b83daf473))

### Ci

- Added codecov action to build workflow ([df5c464](df5c4640b5106947e8aaeb5115eb8d2ce682da4f))

## [0.2.1] - 2022-04-10

[4506df2](4506df29824a91d56c66841c1f26c2be79abc04e)...[29d0bf2](29d0bf24638df1ba53bea75b121f390190e6d431)

### Performance

- Avoid looking up field type and redis field twice for numeric lookups ([728dbd0](728dbd025f2860e7c56721d96c0f83572436bce8))

## [0.2.0] - 2022-04-10

[0a3d1c8](0a3d1c8bba45ddbc9e082d2f58b90136da5471ee)...[4506df2](4506df29824a91d56c66841c1f26c2be79abc04e)

### Breaking changes

- Map Django DateTimeField to NumericField in Redis ([01d9ced](01d9cedb421bfd02475fe08cc3192ea0dcdf07ac))

### Bug Fixes

- Removed print statement when performing a search ([5f21c7a](5f21c7ac343913984b3c67872a7da1314a55f3fb))
- Use NumericField instead of TextField for object IDs ([dc94208](dc942080b192ea66c0102e677d18c6a2efe4215a))

### Features

- Added support for GreaterThan, GreaterThanOrEqual, LessThan, LessThanOrEqual lookups ([7f11e2a](7f11e2a51815fa50732eeba3eb3066471b859ee5))

### Miscellaneous Tasks

- Added rule to identify breaking changes in git cliff config ([ae49df3](ae49df37eb34369baf0989f3f1c349a52057e11c))

### Testing

- Added tests for get_model_root ([2f2b01a](2f2b01a1ffa54fd6440695874a84c72e2492002a))
- Added test for build_filters ([3225fb9](3225fb9c7fb33a9ec06a3f04afa935340a73c912))
- Added test cases to test_build_filters ([c3ed3be](c3ed3beca567a9372af6868eb72b15941c4aab51))

## [0.1.1] - 2022-04-09

[c786930](c7869309a6f88a6ee3698d8118f0948a4ab8c2ba)...[0a3d1c8](0a3d1c8bba45ddbc9e082d2f58b90136da5471ee)

### Documentation

- Updated README with usage instructions ([e771828](e771828734ec949cd8110074cd502634eb5096d5))

### Miscellaneous Tasks

- Updated pyproject.toml with relevant metadata ([76cffaa](76cffaabc0f898c5d250c74ef7f70606082a9088))


wagtail_redisearch/backend.py
from datetime import date, datetime, time
from typing import List, Union
from uuid import UUID

from django.db import models
from django.db.models.lookups import (
    Exact,
    GreaterThan,
    GreaterThanOrEqual,
    LessThan,
    LessThanOrEqual,
    Lookup,
)
from django.db.models.sql.where import SubqueryConstraint, WhereNode

from wagtail.search.backends.base import (
    BaseSearchBackend,
    BaseSearchResults,
    EmptySearchResults,
    FilterError,
    OrderByFieldError,
    SearchFieldError,
)
from wagtail.search.index import (
    AutocompleteField,
    FilterField,
    RelatedFields,
    SearchField,
    class_is_indexed,
)
from wagtail.search.query import And, Boost, MatchAll, Not, Or, Phrase, PlainText

import redis
from redis import Redis
from redis.commands.search.field import NumericField, TagField, TextField
from redis.commands.search.indexDefinition import IndexDefinition, IndexType
from redis.commands.search.query import Query
from redis.commands.search.querystring import DistjunctUnion, IntersectNode, UnionNode

WagtailQuery = Union[str, PlainText, Phrase, And, Or, Not, MatchAll]
WagtailSearchFields = Union[List[str], None]


def get_model_root(model):
    """
    This function finds the root model for any given model. The root model is
    the highest concrete model that it descends from. If the model doesn't
    descend from another concrete model then the model is it's own root model so
    it is returned.

    Examples:
    >>> get_model_root(wagtail.core.Page) # doctest: +SKIP
    wagtailcore.Page
    >>> get_model_root(myapp.HomePage) # doctest: +SKIP
    wagtailcore.Page
    >>> get_model_root(wagtailimages.Image) # doctest: +SKIP
    wagtailimages.Image
    """
    if model._meta.parents:
        parent_model = list(model._meta.parents.items())[0][0]
        return get_model_root(parent_model)

    return model


redis_field_from_type = {
    "TextField": TextField,
    "CharField": TextField,
    "EmailField": TextField,
    "SlugField": TextField,
    "URLField": TextField,
    "AutoField": NumericField,
    "SmallAutoField": NumericField,
    "BigAutoField": NumericField,
    "BooleanField": TagField,
    "FloatField": NumericField,
    "IntegerField": NumericField,
    "BigIntegerField": NumericField,
    "PositiveBigIntegerField": NumericField,
    "PositiveIntegerField": NumericField,
    "PositiveSmallIntegerField": NumericField,
    "SmallIntegerField": NumericField,
    "DateField": NumericField,
    "DateTimeField": NumericField,
    "TimeField": NumericField,
    "UUIDField": TextField,
}


def get_redis_field(
    model, field, sortable=False, no_index=False, field_prefix=""
) -> Union[TextField, NumericField]:
    field_name = f"{field_prefix}{field.field_name}"
    field_type = field.get_type(model)
    redis_field = redis_field_from_type.get(field_type, None)
    if redis_field is TextField:
        weight = 1
        try:
            if field.boost is not None:
                weight = field.boost
        except AttributeError:
            pass
        return TextField(
            field_name,
            weight=weight,
            sortable=sortable,
            no_index=no_index,
        )
    elif redis_field is NumericField:
        return NumericField(
            field_name,
            sortable=sortable,
            no_index=no_index,
        )
    elif redis_field is TagField:
        return TagField(field_name)
    else:
        raise SearchFieldError(
            f"Field type {field_type} is not yet mapped in wagtail-redisearch"
        )


def value_to_redis(value) -> Union[str, int]:
    """
    Convert value to a redis-compatible type
    >>> value_to_redis(1)
    1
    >>> value_to_redis(1.0)
    1.0
    >>> value_to_redis(True)
    1
    >>> value_to_redis("wagtail")
    'wagtail'
    >>> value_to_redis(["1", "2", "3"])
    '1, 2, 3'
    >>> value_to_redis([1, 2, 3])
    '1, 2, 3'
    >>> from datetime import timezone
    >>> value_to_redis(datetime(2020, 1, 1, tzinfo=timezone.utc))
    1577836800.0
    >>> value_to_redis(date(2020, 1, 1))
    1577836800.0
    >>> value_to_redis(time(11, 55, tzinfo=timezone.utc))
    '115500'
    >>> value_to_redis(UUID("5343fec8-de40-47ff-ac3b-65374f87dc61"))
    '5343fec8-de40-47ff-ac3b-65374f87dc61'
    >>> value_to_redis(None)
    ''
    """
    if isinstance(value, str):
        return value
    elif isinstance(value, bool):
        return int(value)
    elif isinstance(value, int):
        return value
    elif isinstance(value, float):
        return value
    elif isinstance(value, list):
        try:
            return ", ".join(value)
        except TypeError:
            return ", ".join([str(v) for v in value])
    elif isinstance(value, time):
        return f"{value.hour:02d}{value.minute:02d}{value.second:02d}"
    elif isinstance(value, datetime):
        return value.timestamp()
    elif isinstance(value, date):
        return datetime(value.year, value.month, value.day).timestamp()
    elif isinstance(value, UUID):
        return str(value)
    elif value is None:
        return ""
    else:
        raise ValueError(f"Cannot convert type {str(type(value))} to Redis type")


class RediSearchModelIndex:
    def __init__(self, backend, name: str, model):
        self.backend = backend
        self.name = name
        self.client: Redis = backend.client
        self.model = model
        self.fields = model.search_fields

    @property
    def index_prefix(self):
        return f"{self.name}:"

    def document_key(self, document_id):
        return f"{self.index_prefix}{document_id}"

    @property
    def ft(self):
        return self.client.ft(self.name)

    def create(self):
        mapping = {"wagtail_id": NumericField("wagtail_id")}
        try:
            self.ft.create_index(
                fields=mapping.values(),
                definition=IndexDefinition(
                    prefix=[self.index_prefix], index_type=IndexType.HASH
                ),
            )
            return self
        except redis.exceptions.ResponseError as e:
            if str(e) == "Index already exists":
                return self
            else:
                raise e

    def delete(self):
        try:
            self.ft.dropindex(delete_documents=True)
        except redis.exceptions.ResponseError as e:
            # if the index doesn't exist continue
            if str(e) == "Unknown Index name":
                pass

    def add_model(self, model):
        search_fields = model.get_search_fields()

        for field in search_fields:
            if isinstance(field, SearchField) or isinstance(field, AutocompleteField):
                redis_field = get_redis_field(model, field)
                try:
                    self.ft.alter_schema_add(fields=[redis_field])
                except redis.exceptions.ResponseError as e:
                    if "Duplicate field in schema" in str(e):
                        pass
            elif isinstance(field, FilterField):
                info = self.ft.info()
                field_name_b = field.field_name.encode()
                field_exists = any(
                    [attr for attr in info["attributes"] if attr[1] == field_name_b]
                )
                if field_exists:
                    continue
                else:
                    redis_field = TagField(field.field_name)
                    self.ft.alter_schema_add(fields=[redis_field])
            elif isinstance(field, RelatedFields):
                for sub_field in field.fields:
                    redis_field = get_redis_field(
                        model, sub_field, field_prefix=f"{field.field_name}__"
                    )
                    try:
                        self.ft.alter_schema_add(fields=[redis_field])
                    except redis.exceptions.ResponseError as e:
                        if "Duplicate field in schema" in str(e):
                            pass

    def add_item(self, item):
        # Make sure the object can be indexed
        if not class_is_indexed(item.__class__):
            return

        mapping = {"wagtail_id": item.id}
        for field in item.search_fields:
            if (
                isinstance(field, SearchField)
                or isinstance(field, FilterField)
                or isinstance(field, AutocompleteField)
            ):
                value = field.get_value(item)
                mapping[field.field_name] = value_to_redis(value)
            elif isinstance(field, RelatedFields):
                related = field.get_value(item)
                for sub_field in field.fields:
                    field_name = f"{field.field_name}__{sub_field.field_name}"
                    try:
                        value = sub_field.get_value(related)
                    except AttributeError:
                        # this may happen when a foreign key allows nulls
                        value = None
                    mapping[field_name] = value_to_redis(value)
        self.client.hset(self.document_key(item.id), mapping=mapping)

    def add_items(self, model, items):
        if not class_is_indexed(model):
            return
        for item in items:
            self.add_item(item)

    def delete_item(self, item):
        # Make sure the object can be indexed
        if not class_is_indexed(item.__class__):
            return
        self.client.delete(self.document_key(item.id))

    def reset(self):
        """Deletes index and re-creates it"""
        self.delete()
        self.create()


class RediSearchIndexRebuilder:
    index: RediSearchModelIndex

    def __init__(self, index):
        self.index = index

    def start(self):
        return self.index.create()

    def finish(self):
        pass


def build_query_string_autocomplete(
    query: PlainText, queryset: models.QuerySet, fields: WagtailSearchFields
) -> str:
    # if no fields are explicitly provided, search all autocomplete fields
    if fields is None:
        fields = [a.field_name for a in queryset.model.get_autocomplete_search_fields()]
    query_string = (
        f"@{build_filter_fields(queryset, fields)}:{query.query_string.rstrip()}*"
    )
    return query_string


def build_filter_fields(queryset: models.QuerySet, fields: List[str]) -> str:
    """
    Returns union of fields to restrict search to those fields.
    Related fields are expanded to search against all sub-fields.

    >>> build_filter_fields(queryset, ["title", "search_description"]) # doctest: +SKIP
    'title|search_description'
    """

    # expand related fields
    expanded_fields = []
    field_map = {f.field_name: f for f in queryset.model.search_fields}
    for field in fields:
        model_field = field_map[field]
        if isinstance(model_field, RelatedFields):
            for sub_field in model_field.fields:
                expanded_fields.append(f"{field}__{sub_field.field_name}")
        else:
            expanded_fields.append(field)

    return UnionNode(*expanded_fields).to_string(with_parens=False)


def build_query_string(
    query: WagtailQuery,
    queryset: models.QuerySet,
    fields: WagtailSearchFields,
    autocomplete=False,
) -> str:
    """
    Builds a query that conforms to the RediSearch query syntax.
    https://redis.io/docs/stack/search/reference/query_syntax/
    """
    if isinstance(query, PlainText):
        if autocomplete:
            return build_query_string_autocomplete(query, queryset, fields)

        if query.operator == "or":
            query_string = UnionNode(*query.query_string.split()).to_string()
        else:
            query_string = query.query_string
        if fields is None:
            return query_string
        return f"@{build_filter_fields(queryset, fields)}:{query_string}"

    elif isinstance(query, Phrase):
        if fields is None:
            return f'"{query.query_string}"'
        else:
            return f'@{build_filter_fields(queryset, fields)}:"{query.query_string}"'

    elif isinstance(query, And):
        return IntersectNode(
            build_query_string(query.subqueries[0], queryset, fields),
            build_query_string(query.subqueries[1], queryset, fields),
        ).to_string()

    elif isinstance(query, Or):
        return UnionNode(
            build_query_string(query.subqueries[0], queryset, fields),
            build_query_string(query.subqueries[1], queryset, fields),
        ).to_string()

    elif isinstance(query, Boost):
        query_string = build_query_string(query.subquery, queryset, fields)
        return f"{query_string} => {{ $weight: {query.boost} }}"

    elif isinstance(query, Not):
        query_string = build_query_string(query.subquery, queryset, fields)
        return DistjunctUnion(query_string).to_string()

    else:
        raise NotImplementedError(
            "`%s` is not supported by the RediSearch search backend."
            % query.__class__.__name__
        )


def check_fields(
    fields: WagtailSearchFields, queryset: models.QuerySet, autocomplete: bool
):
    """Checks if fields are currently indexed, otherwise raises a SearchFieldError"""
    if fields is None:
        return

    if autocomplete:
        searchable_fields = queryset.model.get_autocomplete_search_fields()
        error_msg = "index.AutocompleteField"
    else:
        searchable_fields = queryset.model.get_search_fields()
        error_msg = "index.SearchField"

    allowed_fields = {field.field_name for field in searchable_fields}

    for field_name in fields:
        if field_name not in allowed_fields:
            raise SearchFieldError(
                f'Cannot search with field "{field_name}". Please add '
                f"{error_msg}('{field_name}') to {queryset.model.__name__}.search_fields.",
                field_name=field_name,
            )


def get_filterable_field(queryset: models.QuerySet, field_name: str):
    return dict(
        (field.get_attname(queryset.model), field)
        for field in queryset.model.get_filterable_search_fields()
    ).get(field_name, None)


def build_sort_by(
    query: Query, queryset: models.QuerySet, order_by_relevance=True
) -> Query:
    if order_by_relevance:
        return query

    for field_name in queryset.query.order_by:
        ascending = False

        if field_name.startswith("-"):
            ascending = True
            field_name = field_name[1:]

        field = get_filterable_field(queryset, field_name)

        if field is None:
            raise OrderByFieldError(
                f'Cannot sort search results with field "{field_name}". Please add '
                f'index.FilterField("{field_name}") to {queryset.model.__name__}.search_fields.',
                field_name=field_name,
            )
        query = query.sort_by(field_name, ascending)
    return query


def build_filters(lookup: Lookup, negated=False) -> str:
    if isinstance(lookup, WhereNode):
        children = []
        for child in lookup.children:
            try:
                child_negated = child.negated
            except AttributeError:
                child_negated = False
            children.append(build_filters(child, negated=child_negated or negated))

        connector = lookup.connector
        if connector == "AND":
            return " ".join(children)
        elif connector == "OR":
            return UnionNode(*children).to_string()
        else:
            raise NotImplementedError(f"Connector {connector} is not supported yet")

    field: models.Field = lookup.lhs.field
    field_name = field.name
    if field_name == "id":
        field_name = "wagtail_id"

    field_type = field.get_internal_type()
    redis_field = redis_field_from_type.get(field_type)
    if redis_field is None:
        raise NotImplementedError(
            f"{field_type} is not yet supported by wagtail-redisearch"
        )

    value = value_to_redis(lookup.rhs)

    if isinstance(lookup, Exact):
        # for NumericField specify value as both min and max
        if redis_field == NumericField:
            value = f"[{value} {value}]"
        # exact searches shouldn't perform a search, so wrap the value in curly braces
        else:
            value = f"{{{value}}}"

    elif isinstance(lookup, GreaterThan):
        if redis_field != NumericField:
            raise FilterError(f"{field_type} is not a numeric field")
        value = f"[({value} inf]"

    elif isinstance(lookup, GreaterThanOrEqual):
        if redis_field != NumericField:
            raise FilterError(f"{field_type} is not a numeric field")
        value = f"[{value} inf]"

    elif isinstance(lookup, LessThan):
        if redis_field != NumericField:
            raise FilterError(f"{field_type} is not a numeric field")
        value = f"[-inf ({value}]"

    elif isinstance(lookup, LessThanOrEqual):
        if redis_field != NumericField:
            raise FilterError(f"{field_type} is not a numeric field")
        value = f"[-inf {value}]"

    elif isinstance(lookup, SubqueryConstraint):
        raise FilterError(
            "Could not apply filter on search results: Subqueries are not allowed."
        )
    else:
        raise FilterError(
            f"Could not apply filter on search results: Unknown where node: {str(type(lookup))}"
        )

    if negated:
        return f"-@{field_name}:{value}"
    else:
        return f"@{field_name}:{value}"


def build_query(
    query: WagtailQuery,
    fields: WagtailSearchFields,
    queryset: models.QuerySet,
    order_by_relevance=False,
    autocomplete=False,
) -> Query:
    """Builds query for RediSearch"""

    check_fields(fields, queryset, autocomplete)

    redis_query = Query(
        " ".join(
            [
                build_query_string(query, queryset, fields, autocomplete),
                build_filters(queryset.query.where),
            ]
        )
    )
    redis_query = build_sort_by(redis_query, queryset, order_by_relevance)
    return redis_query


class RediSearchResults(BaseSearchResults):
    def __init__(
        self,
        query: WagtailQuery,
        queryset: models.QuerySet,
        fields: WagtailSearchFields,
        index: RediSearchModelIndex,
        order_by_relevance: bool,
        autocomplete: bool,
    ):
        self.query = query
        self.queryset = queryset
        self.fields = fields
        self.index = index
        self.order_by_relevance = order_by_relevance
        self.autocomplete = autocomplete
        self.start = 0
        self.stop = None
        self._results_cache = None
        self._count_cache = None
        self._score_field = None

    def _clone(self):
        klass = self.__class__
        new = klass(
            self.query,
            self.queryset,
            self.fields,
            self.index,
            self.order_by_relevance,
            self.autocomplete,
        )
        new.start = self.start
        new.stop = self.stop
        new._score_field = self._score_field
        return new

    def _do_search(self):
        query = build_query(
            self.query,
            self.fields,
            self.queryset,
            self.order_by_relevance,
            self.autocomplete,
        )
        results = self.index.ft.search(query)
        return self.queryset.filter(id__in=[doc.wagtail_id for doc in results.docs])

    def _do_count(self):
        return self._do_search().count()


class RediSearchBackend(BaseSearchBackend):
    index_name = "wagtail"
    rebuilder_class = RediSearchIndexRebuilder
    client: Redis
    DEFAULT_OPERATOR = "and"

    def __init__(self, params):
        super(RediSearchBackend, self).__init__(params)

        self.index_name = params.pop("INDEX", "wagtail")

        host = params.pop("HOST", "127.0.0.1")
        port = params.pop("PORT", 6379)
        self.client = Redis(host=host, port=port, db=0, **params)

    def get_index_for_model(self, model):
        # Split models up into separate indices based on their root model.
        # For example, all page-derived models get put together in one index,
        # while images and documents each have their own index.
        root_model = get_model_root(model)
        index_name = (
            f"{self.index_name}__{root_model._meta.app_label}_{root_model.__name__}"
        )
        return RediSearchModelIndex(self, index_name, model)

    def search(
        self,
        query: Union[str, WagtailQuery],
        model_or_queryset: models.QuerySet,
        fields=None,
        operator=None,
        order_by_relevance=True,
        # partial matching in search() will be completely removed in a future release
        # https://docs.wagtail.org/en/stable/releases/2.15.html#search-method-partial-match-future-deprecation
        partial_match=False,
        autocomplete=False,
    ):
        # Find model/queryset
        if isinstance(model_or_queryset, models.QuerySet):
            model = model_or_queryset.model
            queryset = model_or_queryset
        else:
            model = model_or_queryset
            queryset = model_or_queryset.objects.all()

        # Model must be a class that is in the index
        if not class_is_indexed(model):
            return EmptySearchResults()

        # Check that there's still a query string after the clean up
        if query == "":
            return EmptySearchResults()

        if isinstance(query, str):
            if not autocomplete:
                # prevent users from (ab)using globbing and fuzzy matching
                query = query.replace("*", "").replace("%", "")
            query = PlainText(query, operator=operator or self.DEFAULT_OPERATOR)

        return RediSearchResults(
            query,
            queryset,
            fields,
            self.get_index_for_model(model),
            order_by_relevance,
            autocomplete,
        )

    def autocomplete(self, *args, **kwargs):
        return self.search(*args, **kwargs, autocomplete=True)


SearchBackend = RediSearchBackend
